agent/engine/execcmdagent/manager_linux.go

```diff
@@ -0,0 +1,209 @@
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package execcmdagent
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/aws/amazon-ecs-agent/agent/utils/retry"
+)
+
+// RestartIfStopped restarts the ExecCommandAgent in the container passed as parameter, only for ExecCommandAgent-enabled tasks.
+// The status of the ExecCommandAgent in the container is retrieved using a docker exec inspect call, using the dockerExecID
+// stored in container.GetExecCommandAgentMetadata().DockerExecID.
+//
+// If container.GetExecCommandAgentMetadata() is nil, or the ExecCommandAgent is still running, no action is taken.
+// To actually restart the ExecCommandAgent, this function invokes this instance's Start method.
+func (m *manager) RestartIfStopped(ctx context.Context, client dockerapi.DockerClient, task *apitask.Task, container *apicontainer.Container, containerId string) (bool, error) {
+	if !task.IsExecCommandAgentEnabled() {
+		return false, nil
+	}
+	execMD := container.GetExecCommandAgentMetadata()
+	if execMD == nil { // This means m.Start has never been invoked
+		return false, nil
+	}
+	resp, err := client.InspectContainerExec(ctx, execMD.DockerExecID, dockerclient.ContainerExecInspectTimeout)
+	if err != nil {
+		if _, ok := err.(*dockerapi.DockerTimeoutError); ok {
+			return false, err
+		}
+		// TODO: [ecs-exec] consider retry/adding an extra docker exec inspect API call to get more info
+		return false, err
+	}
+	if resp.Running { // Exec cmd agent is running, nothing to do
+		return false, nil
+	}
+	// Restart if not running
+	//TODO: [ecs-exec] retry only for certain exit codes?
+	seelog.Warnf("Task engine [%s]: ExecCommandAgent Process stopped (exitCode:=%d) for container %s, restarting...", task.Arn, resp.ExitCode, containerId)
+	err = m.Start(ctx, client, task, container, containerId)
+	// err == nil means start succeeded
+	return err == nil, err
+}
+
+// Start idempotently starts the ExecCommandAgent in the container passed as parameter, only for ExecCommandAgent-enabled tasks.
+// In order for the ExecCommandAgent to be successfully started, the following conditions need to be met:
+//	1) There shall be at most one ExecCommandAgent process running in the container (with the name defined in apitask.ExecCommandAgentBinName)
```

Comment:
why is this check a requirement? ie. when do we anticipate a process already running?

Reply:
As per the design doc, there might be customers already running their own ssm agent in their containers. 
For these cases we don't want to start another ssm agent.

Reply:
Removed that check altogether... less code, yay!